### PR TITLE
[log232] Create PrometheusRule object with elasticsearch rules and alerts

### PR DESCRIPTION
**Pull request by josefkarasek**
Rules & alerts definitions are read from two files, that need to exist:
* files/prometheus_alerts.yml
* files/prometheus_rules.yml

These two files are generated by github.com/lukas-vlcek/elasticsearch-mixin

### CODE REVIEW

**Comment by josefkarasek** [comment]
@ewolinetz I'm vendoring in monitoring client - that's why this PR looks so huge.
This client is needed for `PrometheusRules` and `ServiceMonitor` objects
`github.com/coreos/prometheus-operator/pkg/client/monitoring/v1`

File: pkg/k8shandler/prometheus_rule.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sYAML "k8s.io/apimachinery/pkg/util/yaml"
+)
+
+const (
+	alertsFilePath = "files/prometheus_alerts.yml"
+	rulesFilePath  = "files/prometheus_rules.yml"
+)
+
+func CreateOrUpdatePrometheusRules(dpl *v1alpha1.Elasticsearch) error {
```
Comment (ewolinetz) [inline pkg/k8shandler/prometheus_rule.go]: I don't see a reason to have both this available at both scopes. We can likely just use `CreateOrUpdatePrometheusRules` and move the logic from within `createOrUpdatePrometheusRules` into this function.

File: pkg/k8shandler/prometheus_rule.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	alertsRuleSpec.Groups = append(alertsRuleSpec.Groups, rulesRuleSpec.Groups...)
+
+	rule := monitoringv1.PrometheusRule{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       monitoringv1.PrometheusRuleKind,
+			APIVersion: monitoringv1.Group + "/" + monitoringv1.Version,
```
Comment (ewolinetz) [inline pkg/k8shandler/prometheus_rule.go]: `monitoringv1.SchemeGroupVersion.String()` instead

Comment (josefkarasek) [reply pkg/k8shandler/prometheus_rule.go]: Where did you find it?
`monitoringv1` only defines `Group` and `Version`

**Comment by josefkarasek** [comment]
@ewolinetz @lukas-vlcek @brancz
Sample output, does it look OK?
```yaml
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
metadata:
  creationTimestamp: 2018-11-01T11:40:15Z
  generation: 1
  labels:
    component: es
  name: elasticsearch-prometheus-rules
  namespace: openshift-logging
  ownerReferences:
  - apiVersion: logging.openshift.io/v1alpha1
    controller: true
    kind: Elasticsearch
    name: elasticsearch
    uid: 42a66e1e-ddc6-11e8-b0d5-0e4ec0971078
  resourceVersion: "17329"
  selfLink: /apis/monitoring.coreos.com/v1/namespaces/openshift-logging/prometheusrules/elasticsearch-prometheus-rules
  uid: e699c1f4-ddca-11e8-b0d5-0e4ec0971078
spec:
  groups:
  - name: logging_elasticsearch.alerts
    rules:
    - alert: ElasticsearchClusterNotHealthy
      annotations:
        description: Cluster {{ $labels.cluster }} health status has been RED for
          at least 2m. Cluster does not accept writes, shards may be missing or master
          node hasn't been elected yet.
        summary: Cluster health status is RED
      expr: |
        sum by (cluster) (es_cluster_status == 2)
      for: 2m
      labels:
        severity: critical
    - alert: ElasticsearchClusterNotHealthy
      annotations:
        description: Cluster {{ $labels.cluster }} health status has been YELLOW for
          at least 20m. Some shard replicas are not allocated.
        summary: Cluster health status is YELLOW
      expr: |
        sum by (cluster) (es_cluster_status == 1)
      for: 20m
      labels:
        severity: warning
    - alert: ElasticsearchBulkRequestsRejectionJumps
      annotations:
        description: High Bulk Rejection Ratio at {{ $labels.node }} node in {{ $labels.cluster
          }} cluster. This node may not be keeping up with the indexing speed.
        summary: High Bulk Rejection Ratio - {{ $value }}%
      expr: |
        round( bulk:reject_ratio:rate2m * 100, 0.001 ) > 5
      for: 10m
      labels:
        severity: warning
    - alert: ElasticsearchNodeDiskWatermarkReached
      annotations:
        description: Disk Low Watermark Reached at {{ $labels.node }} node in {{ $labels.cluster
          }} cluster. Shards can not be allocated to this node anymore. You should
          consider adding more disk to the node.
        summary: Disk Low Watermark Reached - disk saturation is {{ $value }}%
      expr: |
        sum by (cluster, instance, node) (
          round(
            (1 - (
              es_fs_path_available_bytes /
              es_fs_path_total_bytes
            )
          ) * 100, 0.001)
        ) > 85
      for: 5m
      labels:
        severity: alert
    - alert: ElasticsearchNodeDiskWatermarkReached
      annotations:
        description: Disk High Watermark Reached at {{ $labels.node }} node in {{
          $labels.cluster }} cluster. Some shards will be re-allocated to different
          nodes if possible. Make sure more disk space is added to the node or drop
          old indices allocated to this node.
        summary: Disk High Watermark Reached - disk saturation is {{ $value }}%
      expr: |
        sum by (cluster, instance, node) (
          round(
            (1 - (
              es_fs_path_available_bytes /
              es_fs_path_total_bytes
            )
          ) * 100, 0.001)
        ) > 90
      for: 5m
      labels:
        severity: high
    - alert: ElasticsearchNodeDiskLowForSegmentMerges
      annotations:
        description: Free disk at {{ $labels.node }} node in {{ $labels.cluster }}
          cluster may be low for optimal segment merges
        summary: Free disk may be low for optimal segment merges
      expr: |
        sum by (cluster, instance, node) (es_fs_path_free_bytes) /
        sum by (cluster, instance, node) (es_indices_store_size_bytes)
        < 0.5
      for: 15m
      labels:
        severity: warning
    - alert: ElasticsearchJVMHeapUseHigh
      annotations:
        description: JVM Heap usage on the node {{ $labels.node }} in {{ $labels.cluster
          }} cluster is {{ $value }}%.
        summary: JVM Heap usage on the node is high
      expr: |
        sum by (cluster, instance, node) (es_jvm_mem_heap_used_percent) > 75
      for: 10m
      labels:
        severity: alert
    - alert: AggregatedLoggingSystemCPUHigh
      annotations:
        description: System CPU usage on the node {{ $labels.node }} in {{ $labels.cluster
          }} cluster is {{ $value }}%
        summary: System CPU usage is high
      expr: |
        sum by (cluster, instance, node) (es_os_cpu_percent) > 90
      for: 1m
      labels:
        severity: alert
    - alert: ElasticsearchProcessCPUHigh
      annotations:
        description: ES process CPU usage on the node {{ $labels.node }} in {{ $labels.cluster
          }} cluster is {{ $value }}%
        summary: ES process CPU usage is high
      expr: |
        sum by (cluster, instance, node) (es_process_cpu_percent) > 90
      for: 1m
      labels:
        severity: alert
  - name: logging_elasticsearch.rules
    rules:
    - expr: |
        rate(es_threadpool_threads_count{name="bulk", type="rejected"}[2m])
      record: bulk:rejected_requests:rate2m
    - expr: |
        rate(es_threadpool_threads_count{name="bulk", type="completed"}[2m])
      record: bulk:completed_requests:rate2m
    - expr: |
        sum by (cluster, instance, node) (bulk:rejected_requests:rate2m) / on (cluster, instance, node) (bulk:completed_requests:rate2m)
      record: bulk:reject_ratio:rate2m
```

**Comment by brancz** [comment]
That generally looks good to me.

File: pkg/k8shandler/prometheus_rule.go (hunk 1 above)
Comment (ewolinetz) [reply pkg/k8shandler/prometheus_rule.go]: We don't get anything from having the private implementation though, we never call it again anywhere else. To me its just making the code more difficult to read without adding any DRY

File: pkg/k8shandler/prometheus_rule.go (hunk 2 above)
Comment (ewolinetz) [reply pkg/k8shandler/prometheus_rule.go]: https://github.com/openshift/elasticsearch-operator/blob/master/pkg/apis/elasticsearch/v1alpha1/register.go#L20

**Pull request merged**